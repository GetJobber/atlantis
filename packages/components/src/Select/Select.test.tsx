import React from "react";
import renderer from "react-test-renderer";
import { cleanup, fireEvent, render } from "@testing-library/react";
import { Option, Select } from ".";

afterEach(cleanup);

it("renders a Select with no options", () => {
  const tree = renderer.create(<Select />).toJSON();
  expect(tree).toMatchSnapshot();
});

// It seems dumb to test this, but the children type is
// ReactNode | ReactNode[] and it complains that ReactNode[] needs more than one
// element if you try to use just that as the type.
it("renders a Select with one option", () => {
  const tree = renderer
    .create(
      <Select>
        <Option>Foo</Option>
      </Select>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders a Select with many options", () => {
  const tree = renderer
    .create(
      <Select>
        <Option>Foo</Option>
        <Option>Bar</Option>
        <Option>Baz</Option>
        <Option>Quux</Option>
      </Select>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly as small", () => {
  const tree = renderer
    .create(
      <Select size="small">
        <Option>Foo</Option>
      </Select>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly as large", () => {
  const tree = renderer
    .create(
      <Select size="large">
        <Option>Foo</Option>
      </Select>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly when disabled", () => {
  const tree = renderer
    .create(
      <Select disabled>
        <Option>Foo</Option>
      </Select>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly when invalid", () => {
  const tree = renderer
    .create(
      <Select invalid>
        <Option>Foo</Option>
      </Select>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders the defaultValue when set.", () => {
  const { container } = render(
    <Select defaultValue="bar">
      <Option value="foo">Foo</Option>
      <Option value="bar">Bar</Option>
    </Select>,
  );

  const select = container.querySelector("select") as HTMLSelectElement;

  expect(select.options[select.selectedIndex].value).toBe("bar");
});

it("renders the value when set.", () => {
  const { container } = render(
    <Select value="bar">
      <Option value="foo">Foo</Option>
      <Option value="bar">Bar</Option>
    </Select>,
  );

  const select = container.querySelector("select") as HTMLSelectElement;

  expect(select.options[select.selectedIndex].value).toBe("bar");
});

it("should pass the new value to the onChange handler when the selected option changes.", () => {
  const changeHandler = jest.fn();

  const expectedValue = "bar";

  const { container } = render(
    <Select onChange={changeHandler}>
      <Option value="foo">Foo</Option>
      <Option value="bar">Bar</Option>
    </Select>,
  );

  const select = container.querySelector("select") as HTMLSelectElement;

  fireEvent.change(select, {
    target: { value: expectedValue },
  });
  expect(changeHandler).toHaveBeenCalledWith(expectedValue);
});
