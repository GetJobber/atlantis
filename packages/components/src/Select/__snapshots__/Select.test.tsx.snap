// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Select with many options 1`] = `
<div
  className="wrapper"
  style={
    Object {
      "--formField-maxLength": undefined,
    }
  }
>
  <label
    className="label"
    htmlFor="123e4567-e89b-12d3-a456-426655440003"
  >
     
  </label>
  <select
    className="formField select"
    id="123e4567-e89b-12d3-a456-426655440003"
    onChange={[Function]}
  >
    <option>
      Foo
    </option>
    <option>
      Bar
    </option>
    <option>
      Baz
    </option>
    <option>
      Quux
    </option>
  </select>
  <span
    className="postfix"
  >
    <svg
      className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
      data-testid="arrowDown"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        className=""
        d="M328.677 353.767c-8.333-8.309-19.249-12.453-30.16-12.433-10.876 0.019-21.745 4.175-30.042 12.466s-12.456 19.153-12.475 30.021c-0.019 10.918 4.139 21.842 12.475 30.172l213.334 213.51c8.337 8.333 19.264 12.497 30.191 12.497s21.854-4.164 30.191-12.497l213.333-213.33c8.337-8.33 12.497-19.254 12.476-30.172-0.021-10.868-4.177-21.729-12.476-30.021-8.294-8.291-19.166-12.447-30.042-12.466-10.91-0.019-21.828 4.125-30.161 12.433l-183.322 182.979-183.323-183.159z"
      />
    </svg>
  </span>
</div>
`;

exports[`renders a Select with no options 1`] = `
<div
  className="wrapper"
  style={
    Object {
      "--formField-maxLength": undefined,
    }
  }
>
  <label
    className="label"
    htmlFor="123e4567-e89b-12d3-a456-426655440001"
  >
     
  </label>
  <select
    className="formField select"
    id="123e4567-e89b-12d3-a456-426655440001"
    onChange={[Function]}
  />
  <span
    className="postfix"
  >
    <svg
      className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
      data-testid="arrowDown"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        className=""
        d="M328.677 353.767c-8.333-8.309-19.249-12.453-30.16-12.433-10.876 0.019-21.745 4.175-30.042 12.466s-12.456 19.153-12.475 30.021c-0.019 10.918 4.139 21.842 12.475 30.172l213.334 213.51c8.337 8.333 19.264 12.497 30.191 12.497s21.854-4.164 30.191-12.497l213.333-213.33c8.337-8.33 12.497-19.254 12.476-30.172-0.021-10.868-4.177-21.729-12.476-30.021-8.294-8.291-19.166-12.447-30.042-12.466-10.91-0.019-21.828 4.125-30.161 12.433l-183.322 182.979-183.323-183.159z"
      />
    </svg>
  </span>
</div>
`;

exports[`renders a Select with one option 1`] = `
<div
  className="wrapper"
  style={
    Object {
      "--formField-maxLength": undefined,
    }
  }
>
  <label
    className="label"
    htmlFor="123e4567-e89b-12d3-a456-426655440002"
  >
     
  </label>
  <select
    className="formField select"
    id="123e4567-e89b-12d3-a456-426655440002"
    onChange={[Function]}
  >
    <option>
      Foo
    </option>
  </select>
  <span
    className="postfix"
  >
    <svg
      className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
      data-testid="arrowDown"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        className=""
        d="M328.677 353.767c-8.333-8.309-19.249-12.453-30.16-12.433-10.876 0.019-21.745 4.175-30.042 12.466s-12.456 19.153-12.475 30.021c-0.019 10.918 4.139 21.842 12.475 30.172l213.334 213.51c8.337 8.333 19.264 12.497 30.191 12.497s21.854-4.164 30.191-12.497l213.333-213.33c8.337-8.33 12.497-19.254 12.476-30.172-0.021-10.868-4.177-21.729-12.476-30.021-8.294-8.291-19.166-12.447-30.042-12.466-10.91-0.019-21.828 4.125-30.161 12.433l-183.322 182.979-183.323-183.159z"
      />
    </svg>
  </span>
</div>
`;

exports[`renders correctly as large 1`] = `
<div
  className="wrapper large"
  style={
    Object {
      "--formField-maxLength": undefined,
    }
  }
>
  <label
    className="label"
    htmlFor="123e4567-e89b-12d3-a456-426655440005"
  >
     
  </label>
  <select
    className="formField select"
    id="123e4567-e89b-12d3-a456-426655440005"
    onChange={[Function]}
  >
    <option>
      Foo
    </option>
  </select>
  <span
    className="postfix"
  >
    <svg
      className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
      data-testid="arrowDown"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        className=""
        d="M328.677 353.767c-8.333-8.309-19.249-12.453-30.16-12.433-10.876 0.019-21.745 4.175-30.042 12.466s-12.456 19.153-12.475 30.021c-0.019 10.918 4.139 21.842 12.475 30.172l213.334 213.51c8.337 8.333 19.264 12.497 30.191 12.497s21.854-4.164 30.191-12.497l213.333-213.33c8.337-8.33 12.497-19.254 12.476-30.172-0.021-10.868-4.177-21.729-12.476-30.021-8.294-8.291-19.166-12.447-30.042-12.466-10.91-0.019-21.828 4.125-30.161 12.433l-183.322 182.979-183.323-183.159z"
      />
    </svg>
  </span>
</div>
`;

exports[`renders correctly as small 1`] = `
<div
  className="wrapper small"
  style={
    Object {
      "--formField-maxLength": undefined,
    }
  }
>
  <label
    className="label"
    htmlFor="123e4567-e89b-12d3-a456-426655440004"
  >
     
  </label>
  <select
    className="formField select"
    id="123e4567-e89b-12d3-a456-426655440004"
    onChange={[Function]}
  >
    <option>
      Foo
    </option>
  </select>
  <span
    className="postfix"
  >
    <svg
      className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
      data-testid="arrowDown"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        className=""
        d="M328.677 353.767c-8.333-8.309-19.249-12.453-30.16-12.433-10.876 0.019-21.745 4.175-30.042 12.466s-12.456 19.153-12.475 30.021c-0.019 10.918 4.139 21.842 12.475 30.172l213.334 213.51c8.337 8.333 19.264 12.497 30.191 12.497s21.854-4.164 30.191-12.497l213.333-213.33c8.337-8.33 12.497-19.254 12.476-30.172-0.021-10.868-4.177-21.729-12.476-30.021-8.294-8.291-19.166-12.447-30.042-12.466-10.91-0.019-21.828 4.125-30.161 12.433l-183.322 182.979-183.323-183.159z"
      />
    </svg>
  </span>
</div>
`;

exports[`renders correctly when disabled 1`] = `
<div
  className="wrapper disabled"
  style={
    Object {
      "--formField-maxLength": undefined,
    }
  }
>
  <label
    className="label"
    htmlFor="123e4567-e89b-12d3-a456-426655440006"
  >
     
  </label>
  <select
    className="formField select"
    disabled={true}
    id="123e4567-e89b-12d3-a456-426655440006"
    onChange={[Function]}
  >
    <option>
      Foo
    </option>
  </select>
  <span
    className="postfix"
  >
    <svg
      className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
      data-testid="arrowDown"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        className=""
        d="M328.677 353.767c-8.333-8.309-19.249-12.453-30.16-12.433-10.876 0.019-21.745 4.175-30.042 12.466s-12.456 19.153-12.475 30.021c-0.019 10.918 4.139 21.842 12.475 30.172l213.334 213.51c8.337 8.333 19.264 12.497 30.191 12.497s21.854-4.164 30.191-12.497l213.333-213.33c8.337-8.33 12.497-19.254 12.476-30.172-0.021-10.868-4.177-21.729-12.476-30.021-8.294-8.291-19.166-12.447-30.042-12.466-10.91-0.019-21.828 4.125-30.161 12.433l-183.322 182.979-183.323-183.159z"
      />
    </svg>
  </span>
</div>
`;

exports[`renders correctly when invalid 1`] = `
<div
  className="wrapper invalid"
  style={
    Object {
      "--formField-maxLength": undefined,
    }
  }
>
  <label
    className="label"
    htmlFor="123e4567-e89b-12d3-a456-426655440007"
  >
     
  </label>
  <select
    className="formField select"
    id="123e4567-e89b-12d3-a456-426655440007"
    onChange={[Function]}
  >
    <option>
      Foo
    </option>
  </select>
  <span
    className="postfix"
  >
    <svg
      className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
      data-testid="arrowDown"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        className=""
        d="M328.677 353.767c-8.333-8.309-19.249-12.453-30.16-12.433-10.876 0.019-21.745 4.175-30.042 12.466s-12.456 19.153-12.475 30.021c-0.019 10.918 4.139 21.842 12.475 30.172l213.334 213.51c8.337 8.333 19.264 12.497 30.191 12.497s21.854-4.164 30.191-12.497l213.333-213.33c8.337-8.33 12.497-19.254 12.476-30.172-0.021-10.868-4.177-21.729-12.476-30.021-8.294-8.291-19.166-12.447-30.042-12.466-10.91-0.019-21.828 4.125-30.161 12.433l-183.322 182.979-183.323-183.159z"
      />
    </svg>
  </span>
</div>
`;
